lib/execution_segment.go

```diff
@@ -633,52 +610,120 @@ func (et *ExecutionTuple) fillCache() {
+		SegmentIndex: segmentIndex,
+	}
+}
+
+// GetNewExecutionSegmentSequenceFromValue uses the value provided, splits it
+// between all the segments, using the striping offsets in the sequence,
+// generating a new segment sequence. It then returns a new
+// ExecutionSegmentSequenceWrapper, with the new sequence and segments, such
+// that each new segment in the new sequence has length `Scale(value)/value`
+// while keeping the order.
+//
+// Additionally, the position of a given segment index can be tracked (since
+// empty segments are removed), so that you can reconstruct an ExecutionTuple,
+// if required. If the segment with the trackedIndex is not part of the new
+// sequence, or if a new sequence cannot be generated (for example, for 0
+// values), an error will be returned.
+func (essw *ExecutionSegmentSequenceWrapper) GetNewExecutionSegmentSequenceFromValue(value int64, trackedIndex int) (
+	newSequence *ExecutionSegmentSequenceWrapper, newIndex int, err error,
+) {
+	if value < 1 {
+		return nil, -1, fmt.Errorf("cannot generate new sequence for value %d", value)
+	}
+
+	if value%essw.lcd == 0 { // the value is perfectly divisible so we will get the same tuple
+		return essw, trackedIndex, nil
+	}
+
+	newIndex = -1
+	newESS := make(ExecutionSegmentSequence, 0, len(essw.ExecutionSegmentSequence)) // this can be smaller
+
+	prev := int64(0)
+	for i := range essw.ExecutionSegmentSequence {
+		newValue := essw.ScaleInt64(i, value)
 		if newValue == 0 {
 			continue
 		}
-		var currentES = newExecutionSegment(big.NewRat(prev, value), big.NewRat(prev+newValue, value))
+		currentES := newExecutionSegment(big.NewRat(prev, value), big.NewRat(prev+newValue, value))
 		prev += newValue
-		if i == et.esIndex {
-			newES = currentES
-			esIndex = len(newESS)
+		if i == trackedIndex {
+			newIndex = len(newESS)
 		}
 		newESS = append(newESS, currentES)
 	}
-	return &ExecutionTuple{
-		ES:       newES,
-		sequence: newESS,
-		esIndex:  esIndex,
-		once:     new(sync.Once),
+
+	if newIndex == -1 {
+		return nil, -1, fmt.Errorf(
+			"segment %d (%s) isn't present in the new sequence",
+			trackedIndex, essw.ExecutionSegmentSequence[trackedIndex],
+		)
+	}
+
+	return NewExecutionSegmentSequenceWrapper(newESS), newIndex, nil
+}
+
+// ExecutionTuple is the combination of an ExecutionSegmentSequence(Wrapper) and
+// a specific ExecutionSegment from it. It gives easy access to the efficient
+// scaling and striping algorithms for that specific segment, since the results
+// are cached in the sequence wrapper.
+type ExecutionTuple struct { // TODO rename? make fields private and have getter methods?
+	Sequence     *ExecutionSegmentSequenceWrapper
+	Segment      *ExecutionSegment
+	SegmentIndex int
+}
+
+func (et *ExecutionTuple) String() string {
+	return fmt.Sprintf("%s in %s", et.Segment, et.Sequence)
+}
+
+// NewExecutionTuple returns a new ExecutionTuple for the provided segment and
+// sequence.
+//
+// TODO: don't return a pointer?
```

Comment:
The Pointer was required because fillCached needed to change the tuple so ... now this can be a not pointer :D

Reply:
yeah, and I started doing it, but then I had to change like a 1000 uses... I don't mind doing it now, but if I had done it, I imagined you would have said (not incorrectly) something like "this could have been done later" :wink:

Reply:
It probably would - I remember moving from the non-pointer to the pointer variant and it took half an afternoon :). To be honest I don't think this is worth an issue, and it's likely won't better anything if changed at this point. I just wanted to point (pun intented :P) out why  this was the case :D